assignments/hw1/ftregister.go

```diff
@@ -0,0 +1,114 @@
+package ftregister
+
+import (
+	"fmt"
+	"github.com/dati-mipt/distributed-storage-algorithms/network"
+	"github.com/dati-mipt/distributed-storage-algorithms/util"
+)
+
+type FaultTolerantRegister struct {
+	rid      int64
+	current  util.TimestampedValue
+	replicas map[int64]network.Link
+}
+
+func NewFaultTolerantRegister(rid int64) *FaultTolerantRegister {
+	return &FaultTolerantRegister{
+		rid:      rid,
+		replicas: map[int64]network.Link{},
+	}
+}
+
+func (r *FaultTolerantRegister) Write(value int64) bool {
+	fmt.Println("-----------Write-----------")
+	fmt.Printf("write value: %d\n", value)
+
+	r.current.Val = value
+	r.current.Ts = util.Timestamp{Number: r.current.Ts.Number + 1, Rid: r.rid}
+
+	return true
+}
+
+func Callback(rep network.Link, pipe chan util.TimestampedValue) {
+	var msg = (rep.BlockingMessage(nil)).(util.TimestampedValue)
+	fmt.Println("waiting ...")
+	pipe <- msg
+}
+
+func BlockingMessageToQuorum(r *FaultTolerantRegister) (msg util.TimestampedValue) {
+	fmt.Println("BlockingMessageToQuorum ...")
+
+	// create channel
+	message_chan := make(chan util.TimestampedValue, len(r.replicas))
+	// done := make(chan bool) // synchronization
+
+	// create go routines
+	var counter = 0
+	for _, rep := range r.replicas {
+		go func( /*pipe chan util.TimestampedValue, rep network.Link*/ ) {
+
+			var msg = (rep.BlockingMessage(nil)).(util.TimestampedValue)
+			message_chan <- msg
+		}() //(message_chan, rep)
+		counter++
+	}
+
+	// debug info
+	fmt.Printf("counter: %d\n", counter)
+	for i := 0; i < cap(message_chan); i++ {
+		y := <-message_chan
+		fmt.Println(y)
+	}
+	close(message_chan) // ?
+
+	// organize quorum
+	var max_replica_resp = (len(r.replicas) / 2) + 1
+	var rid_ = 0
+	var max_ts util.TimestampedValue
+	for elem := range message_chan {
+		rid_++
+		if elem.Ts.Less(max_ts.Ts) {
+			max_ts = elem
+		}
+	}
+	//
+	if rid_ <= max_replica_resp {
+		return util.TimestampedValue{}
+	}
+
+	// <-done
+	return max_ts
+}
+
+// compound read = {read + write}
+
+func (r *FaultTolerantRegister) Read() int64 {
```

Comment:
Blocking message to quorum should be called 2 times:
1) For reading data from replicas.
2) For writing existing max value on other quorum. This is important to ensure linearizability.